src/HappyConnOpener.cc

```diff
@@ -0,0 +1,875 @@
+        peerConnectFailed(peer);
+    params.conn->close(); // TODO: Comm::ConnOpener should do this instead.
+
+    // remember the last failure (we forward it if we cannot connect anywhere)
+    lastFailedConnection = params.conn;
+    delete lastError;
+    lastError = nullptr; // in case makeError() throws
+    lastError = makeError(ERR_CONNECT_FAIL);
+    lastError->xerrno = params.xerrno;
+
+    if (spareWaiting)
+        updateSpareWaitAfterPrimeFailure();
+
+    checkForNewConnection();
+}
+
+/// reacts to a prime attempt failure
+void
+HappyConnOpener::updateSpareWaitAfterPrimeFailure()
+{
+    Must(currentPeer);
+    Must(!prime);
+    Must(spareWaiting);
+
+    if (destinations->doneWithPrimes(*currentPeer)) {
+        cancelSpareWait("all primes failed");
+        ignoreSpareRestrictions = true;
+        return; // checkForNewConnection() will open a spare connection ASAP
+    }
+
+    if (spareWaiting.toGivePrimeItsChance)
+        stopGivingPrimeItsChance();
+
+    // may still be spareWaiting.forSpareAllowance or
+    // may still be spareWaiting.forPrimesToFail
+}
+
+/// called when the prime attempt has used up its chance for a solo victory
+void
+HappyConnOpener::stopGivingPrimeItsChance() {
+    Must(spareWaiting.toGivePrimeItsChance);
+    spareWaiting.toGivePrimeItsChance = false;
+    ThePrimeChanceGiver.dequeue(*this);
+}
+
+/// called when the spare attempt should no longer obey spare connection limits
+void
+HappyConnOpener::stopWaitingForSpareAllowance() {
+    Must(spareWaiting.forSpareAllowance);
+    spareWaiting.forSpareAllowance = false;
+
+    if (spareWaiting.callback)
+        TheSpareAllowanceGiver.jobDroppedAllowance();
+    TheSpareAllowanceGiver.dequeue(*this); // clears spareWaiting.callback
+}
+
+/// stops waiting for the right conditions to open a spare connection
+void
+HappyConnOpener::cancelSpareWait(const char *reason)
+{
+    debugs(17, 5, "because " << reason);
+    Must(spareWaiting);
+
+    if (spareWaiting.toGivePrimeItsChance)
+        stopGivingPrimeItsChance();
+    else if (spareWaiting.forSpareAllowance)
+        stopWaitingForSpareAllowance();
+
+    spareWaiting.clear();
+}
+
+/** Called when an external event changes initiator interest, destinations,
+ * prime, spare, or spareWaiting. Leaves HappyConnOpener in one of these
+ * (mutually exclusive beyond the exceptional state #0) "stable" states:
+ *
+ * 0. Exceptional termination: done()
+ * 1. Processing a single peer: currentPeer
+ *    1.1. Connecting: prime || spare
+ *    1.2. Waiting for spare gap and/or paths: !prime && !spare
+ * 2. Waiting for a new peer: destinations->empty() && !destinations->destinationsFinalized && !currentPeer
+ * 3. Finished: destinations->empty() && destinations->destinationsFinalized && !currentPeer
+ */
+void
+HappyConnOpener::checkForNewConnection()
+{
+    debugs(17, 7, *destinations);
+
+    // The order of the top-level if-statements below is important.
```

Comment:
> What are we trying to avoid happening here?

We are trying to provide lower conditions/code with the right state (computed in higher `if` statements), and we are trying to avoid checking the same thing many times. Re-arranging these `if` statements is likely to cause bugs, and some of them may be rather subtle.

Unfortunately, I would not be able to enumerate everything that could go wrong. In fact, I probably do not know all the dependencies! I do know that many attempts to rearrange the current code have failed. Partial enumeration would be worse than no enumeration in this case IMO. This complex code coordinates progress of two dependent but usually parallel work threads, with many limits placed on their coordination. It is difficult to modify it correctly and moving conditions around should be done with great care (and there are general comments elsewhere talking about the overall difficulty of writing this code IIRC).

IMO, the existing comment is sufficient to warn the developer of the danger -- the statements below have complex dependencies -- but if you can suggest a better wording, please do so. I just hope it is not going to be "Do not modify this code unless you know what you are doing!" :-).

Reply:
HappyConnOpener has to run and synchronize the following jobs:
    - starts and monitors a prime connection
    - starts and monitors a second connection under some rules
    - check at the same time for new destinations comming asynchronously from peer_select code
    - check for itself limitations (eg timeouts) to be removed before start a new connection
    - check for global limitations (eg spare connections number) to be removed before start new connection.

It is proved a complex project.